nixos/modules/programs/wayland/niri.nix

```diff
@@ -0,0 +1,76 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.programs.niri;
+in
+{
+  options.programs.niri = {
+    enable = lib.mkEnableOption "Niri, a scrollable-tiling Wayland compositor";
+
+    package = lib.mkPackageOption pkgs "niri" { };
+
+    # Recommended by upstream
+    # https://github.com/YaLTeR/niri/wiki/Important-Software#authentication-agent
+    polkitAgent.enable = lib.mkEnableOption "a PolicyKit agent service" // {
+      default = config.security.polkit.enable;
+      defaultText = "config.security.polkit.enable";
+    };
+
+    # Recommended by upstream
+    # https://github.com/YaLTeR/niri/wiki/Important-Software#portals
+    portalPackage = lib.mkPackageOption pkgs "xdg-desktop-portal-gnome" {
+      # https://github.com/NixOS/nixpkgs/pull/348193#discussion_r1798515394
+      extraDescription = "Note: `xdg-desktop-portal-gnome` is required for screencast support";
+    };
+  };
+
+  config = lib.mkIf cfg.enable (
+    lib.mkMerge [
+      {
+        environment.systemPackages = [ cfg.package ];
+
+        services = {
+          displayManager.sessionPackages = [ cfg.package ];
+
+          # Recommended by upstream
+          # https://github.com/YaLTeR/niri/wiki/Important-Software#portals
+          gnome.gnome-keyring.enable = lib.mkDefault true;
+        };
+
+        systemd.packages = [ cfg.package ];
+
+        xdg.portal = {
+          enable = lib.mkDefault true;
+
+          configPackages = [ cfg.package ];
+          extraPortals = [ cfg.portalPackage ];
+        };
+      }
+
+      (lib.mkIf cfg.polkitAgent.enable {
+        systemd = {
+          packages = [ pkgs.libsForQt5.polkit-kde-agent ];
+          user.services.plasma-polkit-agent = {
+            wantedBy = [ "niri.service" ];
```

Comment:
Does this work? As far as i can tell, this service definition is supposed to be `After=plasma-core.target`, but that won't exist unless KDE plasma is enabled. Does this work if KDE plasma is disabled, and the target is missing? Does this work if KDE plasma is enabled, and the target is present but not started when you open a niri session? Does this interfere with the polkit agent on a KDE plasma session, if this module is enabled as well? (the last one seems like obviously no because `wantedBy` shouldn't prevent it from being started in kde)

This seems otherwise okay, if the answers to above questions are all "it works fine".

Reply:
> Does this work?

Seems to

> Does this work if KDE plasma is disabled, and the target is missing?

Yes

> Does this work if KDE plasma is enabled, and the target is present but not started when you open a niri session?

It should

> Does this interfere with the polkit agent on a KDE plasma session, if this module is enabled as well? 

No

Reply:
Doesn't seem to work for me. I get this error:
```
$ systemctl --user status plasma-polkit-agent
× plasma-polkit-agent.service - KDE PolicyKit Authentication Agent
     Loaded: loaded (/etc/systemd/user/plasma-polkit-agent.service; enabled; preset: ignored)
    Drop-In: /nix/store/ps3kvlqma8n07gpi9nm6fmmlzp6lpyk5-user-units/plasma-polkit-agent.service.d
             └─overrides.conf
     Active: failed (Result: signal) since Sun 2024-10-20 10:16:54 IST; 6s ago
 Invocation: 1c9ccc875c984ae88957a661a1fe6ded
    Process: 1869 ExecStart=/nix/store/kz54544macmpaddfv871cq7ia8sl3icf-polkit-kde-agent-1-5.27.11/libexec/polkit-kde-authentication-agent-1 (code=killed, signal=ABRT)
   Main PID: 1869 (code=killed, signal=ABRT)

Oct 20 10:16:54 hpaio systemd[1663]: Failed to start KDE PolicyKit Authentication Agent.
Oct 20 10:16:54 hpaio systemd[1663]: plasma-polkit-agent.service: Scheduled restart job, restart counter is at 5.
Oct 20 10:16:54 hpaio systemd[1663]: plasma-polkit-agent.service: Start request repeated too quickly.
Oct 20 10:16:54 hpaio systemd[1663]: plasma-polkit-agent.service: Failed with result 'signal'.
Oct 20 10:16:54 hpaio systemd[1663]: Failed to start KDE PolicyKit Authentication Agent.
```

Adding `after = [ "niri.service" ]` fixes it, as it's needed for correct ordering. See my comment https://github.com/sodiboo/niri-flake/issues/348#issuecomment-2364304314 in the niri-flake repo.